### PR TITLE
Handle record swapping between Linux and DxOS

### PR DESCRIPTION
**Pull request by pcd1193182**
This change swaps the incoming REDACT and OBJECT_RANGE records if the send stream is being received from Illumos. This change is only applicable for Delphix customers, where the DRR_REDACT record had a different value in earlier releases. This has been tested in conjunction with appstack changes to handle the modified API. Pushing needs to be coordinated with those changes.

### CODE REVIEW

File: include/libzfs_core.h [hunk 1]
```diff
@@ -96,7 +96,7 @@ int lzc_send_resume_redacted(const char *, const char *, int,
 int lzc_receive(const char *, nvlist_t *, const char *, boolean_t, boolean_t,
     int);
 int lzc_receive_resumable(const char *, nvlist_t *, const char *, boolean_t,
-    boolean_t, int);
+    boolean_t, boolean_t, int);
```
Comment (ahrens) [inline include/libzfs_core.h]: I think it would make it easier to coordinate the appstack changes if we added a new function call, rather than changing the existing one.  e.g. `lzc_receive_resumable2()` or `lzc_receive_resumable_illumos()`

File: module/zfs/dmu_recv.c [hunk 2]
```diff
@@ -2136,6 +2137,16 @@ static int
 receive_read_record(dmu_recv_cookie_t *drc)
 {
 	int err;
+	/*
+	 * We should revert the commit that added this logic once we no longer
+	 * support FCR from 5.3 or earlier.
```
Comment (ahrens) [inline module/zfs/dmu_recv.c]: I think we should also add a comment explaining why this is needed / what the `illumos` argument means.  Perhaps the comment should go at the libzfs_core layer?

File: module/zfs/dmu_recv.c [hunk 3]
```diff
@@ -2136,6 +2137,16 @@ static int
...
+	 */
+	if (drc->drc_illumos) {
+		if (drc->drc_rrd->header.drr_type == DRR_REDACT)
+			drc->drc_rrd->header.drr_type = DRR_OBJECT_RANGE;
+		else if (drc->drc_rrd->header.drr_type == DRR_OBJECT_RANGE)
```
Comment (ahrens) [inline module/zfs/dmu_recv.c]: Since the illumos-based Delphix Engine doesn't use `DRR_OBJECT_RANGE`, does it make sense to handle it here?  vs. e.g. asserting/returning an error if we get this record type.  Maybe this is tied up in the question of whether the new flag means "receive from illumos (e.g. smartOS, OmniOS)" (which will have DRR_OBJECT_RANGE soon) vs "receive from illumos-based delphix engine" (which won't).

Comment (pcd1193182) [reply module/zfs/dmu_recv.c]: It means the latter; that's the only context where redaction landed before encryption. We could  assert here in that case instead.

File: include/libzfs_core.h (hunk 1 above)
Comment (ahrens) [reply include/libzfs_core.h]: What did you decide to do about this?

Comment (pcd1193182) [reply include/libzfs_core.h]: I split it so there is an _illumos function that goes into the underlying layer with B_TRUE, and lzc_receive_resumable() just does what it does now, passing B_FALSE to the lower layer.

File: contrib/pyzfs/libzfs_core/_libzfs_core.py [hunk 4]
```diff
@@ -988,7 +988,7 @@ def lzc_channel_program_nosync(
 
 
 def lzc_receive_resumable(
-    snapname, fd, force=False, raw=False, origin=None, props=None
+    snapname, fd, force=False, raw=False, illumos=None, origin=None, props=None
```
Comment (ahrens) [inline contrib/pyzfs/libzfs_core/_libzfs_core.py]: I think you need to revert this change now, and/or add a new lzc_receive_resumable_illumos() to the python wrapper?

File: contrib/pyzfs/libzfs_core/_libzfs_core.py [hunk 5]
```diff
@@ -1057,7 +1057,7 @@ def lzc_receive_resumable(
         props = {}
     nvlist = nvlist_in(props)
     ret = _lib.lzc_receive_resumable(
-        snapname, nvlist, c_origin, force, raw, fd)
+        snapname, nvlist, c_origin, force, raw, illumos, fd)
```
Comment (ahrens) [inline contrib/pyzfs/libzfs_core/_libzfs_core.py]: same here

File: contrib/pyzfs/libzfs_core/_libzfs_core.py (hunk 4 above)
Comment (pcd1193182) [reply contrib/pyzfs/libzfs_core/_libzfs_core.py]: I don't think we need to add a python wrapper for this, since we don't use  the python wrappers and  no one else is going to get this patch.
